### PR TITLE
Adding shims for addStream and legacy callback APIs

### PR DESCRIPTION
**Pull request by youennf**
**Description**
Adding legacy APIs (addStream) and callback support for setRemoteDescription, setLocalDescription, addIceCandidate, createOffer and createAnswer for Safari.

**Purpose**
Support some legacy websites.

### CODE REVIEW

**Comment by youennf** [comment]
Failing test seems to be selenium-based. Not sure whether this is fine or whether something should be done here.

File: src/js/safari/safari_shim.js [hunk 1]
```diff
@@ -43,7 +51,65 @@ var safariShim = {
...
+    var setRemoteDescription = prototype.setRemoteDescription;
+    var addIceCandidate = prototype.addIceCandidate;
+
+    prototype.createOffer = function(successCallback, failureCallback) {
+      var promise = createOffer.apply(this, [arguments[2]]);
+      if (!failureCallback) {
+        return promise;
+      }
+      promise.then(successCallback, failureCallback);
+      return Promise.resolve();
```
Comment (fippo) [inline src/js/safari/safari_shim.js:72]: that doesn't look right. Shouldn't you return the lowercase promise? (also below)

**Comment by fippo** [comment]
travis failure is due to #509

File: src/js/safari/safari_shim.js [hunk 2]
```diff
@@ -43,7 +51,65 @@ var safariShim = {
...
+    }
+    var prototype = RTCPeerConnection.prototype;
+    var createOffer = prototype.createOffer;
+    var createAnswer = prototype.createAnswer;
+    var setLocalDescription = prototype.setLocalDescription;
+    var setRemoteDescription = prototype.setRemoteDescription;
+    var addIceCandidate = prototype.addIceCandidate;
+
+    prototype.createOffer = function(successCallback, failureCallback) {
+      var promise = createOffer.apply(this, [arguments[2]]);
```
Comment (fippo) [inline src/js/safari/safari_shim.js]: I think you need something along the lines of https://github.com/webrtc/adapter/blob/master/src/js/chrome/chrome_shim.js#L297 to pass through the createOffer(offerOptions) stuff here.

File: src/js/safari/safari_shim.js (hunk 1 above)
Comment (youennf) [reply src/js/safari/safari_shim.js:72]: As per https://www.w3.org/TR/webrtc/#legacy-interface-extensions, createOffer step 7says "Return a promise resolved with undefined".
I guess that we should really return void but cannot due to WebIDL limitations, hence the current design.

File: src/js/safari/safari_shim.js (hunk 2 above)
Comment (youennf) [reply src/js/safari/safari_shim.js]: Right, I'll fix that!

File: src/js/safari/safari_shim.js (hunk 1 above)
Comment (fippo) [reply src/js/safari/safari_shim.js:72]: ah... that makes sense!

File: src/js/safari/safari_shim.js (hunk 2 above)
Comment (jan-ivar) [reply src/js/safari/safari_shim.js]: `options` is the first argument in the promise case, and the third argument in the callback case. This passes up the wrong argument in the promise case.

**Review by jan-ivar — CHANGES_REQUESTED** [review]

**Comment by fippo** [comment]
wrt to unit tests what do you think about this using chai and sinon-chai (happy to make a PR when you're done):
```
  describe('createOffer shim', () => {
    describe('options passing', () => {
      let stub;
      beforeEach(() => {
        stub = sinon.stub();
        window.RTCPeerConnection.prototype.createOffer = stub;
        shim.shimCallbacksAPI();
      });

      it('with no arguments', () => {
        var pc = new RTCPeerConnection();
        pc.createOffer();
        expect(stub).to.have.been.calledWith(undefined);
      });

      it('with two callbacks', () => {
        var pc = new RTCPeerConnection();
        pc.createOffer(null, null);
        expect(stub).to.have.been.calledWith(undefined);
      });

      it('with a non-function first argument', () => {
        var pc = new RTCPeerConnection();
        pc.createOffer(1);
        expect(stub).to.have.been.calledWith(1);
      });

      it('with two callbacks and options', () => {
        var pc = new RTCPeerConnection();
        pc.createOffer(null, null, 1);
        expect(stub).to.have.been.calledWith(1);
      });

      it('with two callbacks and two additional arguments', () => {
        var pc = new RTCPeerConnection();
        pc.createOffer(null, null, 1, 2);
        expect(stub).to.have.been.calledWith(1);
      });
    });
  });
```
I've been wanting to add sinon for stubs for a long time...

**Review by fippo — APPROVED** [review]

**Comment by MirkoBonadei** [comment]
lgtm

**Comment by youennf** [comment]
Don't really know chai and sinon-chai. Happy to see how this would look like.

**Pull request merged**